### PR TITLE
Bugfix/owncloud/oauth2#103

### LINKED ISSUE #103: OAuth2 token revokation does not expire sessions
### Steps to reproduce
1. Create a new OAuth2 session:

screen shot 2017-12-01 at 13 16 14

2. Remove the authorized OAuth2 token in webUI `Security` section 

### Expected behaviour
Sessions for this client/user agent are removed/invalidated as well

### Current behaviour
Sessions keep alive, all request are reponsed with 200

screen shot 2017-12-01 at 13 17 02

Set up:
- ownCloud 10.0.4 RC1 (daily) Build:2017-12-01T04:16:36+00:00 d4eab96823baec0061bc221e35b49dce968cbed9

- OAuth 0.2.1

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Related Issue
fixes owncloud/oauth2#103

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: lib/private/User/BasicAuthModule.php [hunk 1]
```diff
@@ -45,17 +47,12 @@ public function auth(IRequest $request) {
 			return null;
 		}
 
-		// check uid and password
-		$user = $this->manager->checkPassword($request->server['PHP_AUTH_USER'], $request->server['PHP_AUTH_PW']);
-		if ($user instanceof IUser) {
-			return $user;
+		// reuse logClientIn because this method
```
Comment (PVince81) [inline lib/private/User/BasicAuthModule.php]: because this method... yes ?

File: lib/private/User/BasicAuthModule.php [hunk 2]
```diff
@@ -66,6 +63,10 @@ public function getUserPassword(IRequest $request) {
 			return '';
 		}
 
+		if ($this->session->getSession()->exists('app_password')) {
+			// TODO: use proper password
+			throw new \Exception('Not implemented yet');
```
Comment (PVince81) [inline lib/private/User/BasicAuthModule.php]: is this ever triggered ?

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -988,21 +988,30 @@ public function updateSessionTokenPassword($password) {
...
-					// the session is bad -> kill it
-					$this->logout();
-					return false;
+		$lastUser = null;
+		try {
+			foreach ($this->getAuthModules(true) as $module) {
+				$user = $module->auth($request);
+				if ($user !== null) {
+					if ($this->isLoggedIn() && $this->getUser()->getUID() !== $user->getUID()) {
+						throw new Exception();
```
Comment (PVince81) [inline lib/private/User/Session.php]: would be nice to have a message, especially if we don't have a proper type.

no more auto-logout of bad sessions ?

File: lib/private/User/BasicAuthModule.php (hunk 2 above)
Comment (DeepDiver1975) [reply lib/private/User/BasicAuthModule.php]: Not at the moment - this is a safety exception which will punch us in the face as soon as we use it accidentally.

Actually the whole method should not yet be called ... we could just throw the exception right away

File: lib/private/User/Session.php (hunk 3 above)
Comment (DeepDiver1975) [reply lib/private/User/Session.php]: these exceptions are basically just to jump out of the loop and do the logout in the catch block - I'm not to happy with this .....

**Review by PVince81 — APPROVED** [review]
ok then 👍

**Comment by PVince81** [comment]
looks good, yes

**Comment by PVince81** [comment]
@SamuAlfageme can you retest auth issue with this ?

**Comment by PVince81** [comment]
CI fails ? :-/

**Comment by DeepDiver1975** [comment]
> CI fails ? :-/

drone has network issues

**Comment by DeepDiver1975** [comment]
jenkins has failing integration tests

**Comment by PVince81** [comment]
the integration tests seem completely unrelated. trashbin ?! 🙈

**Comment by PVince81** [comment]
oh wait... trashbin result says `Failed asserting that 401 matches expected '200'.`.

Either we broke something or the test was passing by chance before

**Comment by DeepDiver1975** [comment]
I better have a look ....

**Comment by DeepDiver1975** [comment]
```
10:27:08     And logging in using web as "user1"                                                                   # FeatureContext::loggingInUsingWebAs()
10:27:08 [Thu Feb  8 10:27:08 2018] ::1:53702 [200]: /index.php/apps/files/
10:27:08 [Thu Feb  8 10:27:08 2018] ::1:53706 [401]: /index.php/apps/files_trashbin/ajax/list.php?dir=%2F
10:27:08     When as "user1" the file with original path "/renamed_shared/shared_file.txt" is restored             # FeatureContext::elementInTrashIsRestored()
10:27:08       Failed asserting that 401 matches expected '200'.
10:27:09     Then as "user1" the file with original path "/renamed_shared/shared_file.txt" does not exist in trash # FeatureContext::elementIsNotInTrashCheckingOriginalPath()
10:27:09     And user "user1" should see following elements                                                        # FeatureContext::checkElementList()
10:27:09       | /renamed_shared/                |
10:27:09       | /renamed_shared/shared_file.txt |

**Comment by DeepDiver1975** [comment]
The requests as executed by the integration tests are not reusing the session cookies - which might have a side effect ...

**Comment by PVince81** [comment]
ajax 😱 

but it should use the session from the web login

**Comment by PVince81** [comment]
need to check how the tests are actually doing this ajax call, maybe it's just using basic auth for the call instead of cookies...

File: lib/private/User/Session.php [hunk 4]
```diff
@@ -302,8 +302,10 @@ public function getLoginName() {
 	 * @return boolean|null
 	 * @throws LoginException
 	 */
-	public function login($uid, $password) {
-		$this->session->regenerateId();
+	public function login($uid, $password, $recreateSession = true) {
+		if ($recreateSession) {
```
Comment (PVince81) [inline lib/private/User/Session.php]: what if the session contained a different user ? will your protection apply itself in this code path ?

Comment (DeepDiver1975) [reply lib/private/User/Session.php]: this is taken care of in different code locations - verifyheaders and verifytoken

**Comment by PVince81** [comment]
Weird trashbin failures now:
```
When as "user0" the folder with original path "/local_storage/tmp/textfile0.txt" is restored:

File previously located at /local_storage/tmp/textfile0.txt is still in the trashbin
Failed asserting that true is false.
```

no more related to the 401 error

File: lib/private/User/BasicAuthModule.php [hunk 5]
```diff
@@ -52,10 +52,13 @@ public function auth(IRequest $request) {
...
-		if (count($users) !== 1) {
-			return null;
+		if (count($users) === 1) {
+			$user = $this->manager->checkPassword($users[0]->getUID(), $request->server['PHP_AUTH_PW']);
+		}
+		if ($user instanceof IUser) {
+			return $user;
 		}
-		return $this->manager->checkPassword($users[0]->getUID(), $request->server['PHP_AUTH_PW']);
+		throw new \Exception('Invalid credentials');
```
Comment (PVince81) [inline lib/private/User/BasicAuthModule.php]: return false ? Throwing could cause other side effects...

Comment (PVince81) [reply lib/private/User/BasicAuthModule.php]: in TokenAuthModule we return null, let's stay consistent

File: lib/private/User/TokenAuthModule.php [hunk 6]
```diff
@@ -91,4 +84,46 @@ public function auth(IRequest $request) {
 	public function getUserPassword(IRequest $request) {
 		return $this->password;
 	}
+
+	/**
+	 * @param IRequest $request
+	 * @return null|IToken
+	 */
+	private function getTokenForAppPassword(IRequest $request, &$token) {
+		if (empty($request->server['PHP_AUTH_USER']) || empty($request->server['PHP_AUTH_PW'])) {
```
Comment (PVince81) [inline lib/private/User/TokenAuthModule.php]: not afraid of "0" users ?

**Review by PVince81 — COMMENTED** [review]
see comments

File: lib/private/User/TokenAuthModule.php (hunk 6 above)
Comment (DeepDiver1975) [reply lib/private/User/TokenAuthModule.php]: @PVince81 as requested: fuck php

File: lib/private/User/BasicAuthModule.php (hunk 5 above)
Comment (DeepDiver1975) [reply lib/private/User/BasicAuthModule.php]: null is returned if there is no authentication header to be evaluated aka handled by this module. 
exception is throws if credentials are wrong
user object is returned if creds are valid

Comment (PVince81) [reply lib/private/User/BasicAuthModule.php]: okay

Comment (DeepDiver1975) [reply lib/private/User/BasicAuthModule.php]: But you are right - this logic seems to be missing in the token auch module ....

**Comment by SamuAlfageme** [comment]
@DeepDiver1975 initial tests suggest the problem no longer reproduces with this + owncloud/oauth2#112; great news! 🎉 

1. Reduced drastically `AccessToken`'s `resetExpires()` to bump `expires` every 20s
2. Authorized the Desktop client 
3. Close the client during `access_token` validity period and wait 20s
4. Replay any successful request the client made previously 
5. Observe 401 "Unauthorized" reply on the replayed request 

Would a traversal unit test like this one make sense to be included in the OAuth2 test pool?

Will focus now on trying to reproduce https://github.com/owncloud/oauth2/issues/103 on the mobile apps; with and without these patches.

**Comment by SamuAlfageme** [comment]
@DeepDiver1975 can you check if the scope of this patch can be extended to the old WebDAV endpoint? 

After some more tests, I believe the original cause of https://github.com/owncloud/oauth2/issues/103 is the logic for the old endpoint. Take these 2 `PROPFIND`s using OAuth2 as the auth8n mechanism:

<p align="center">
  
</p>

- First 2 requests (url ending on `:4443`) belong to a 10.0.6 "unpatched" server (1st. is iOS: old endpoint, 2nd. Desktop, new)
- The following 2 are 10.0.6 + https://github.com/owncloud/core/pull/30365 + https://github.com/owncloud/oauth2/issues/112

Replaying the iOS request any number of times results on 2XX, while desktop gets the correct 401.

**Comment by SamuAlfageme** [comment]
Hm. Forcing the use of the old endpoint on the Desktop client (setting env. `OWNCLOUD_CHUNKING_NG=0`) still results on 401 & token refresh. 

Appart from the endpoints there are a few apparent differences between iOS and Desktop `PROPFINDS` 😕 

### Desktop:

```bash
curl \
    -H 'Depth:0' \
    -H 'Authorization:Bearer VBvulBvRbhOpaA5BLkaLl7H7AdeatO5kwGyDNqfwYf2XRnsJTz0sEagLwgSuT3U3' \
    -H 'User-Agent:Mozilla/5.0 (Macintosh) mirall/2.4.0 (build 8836)' \
    -H 'Accept:*/*' \
    -H 'Content-Type:text/xml; charset=utf-8' \
    -H 'X-Request-ID:83fd38fa-395b-429e-a1be-80aad5fc9c32' \
    -H 'Cookie:oc_sessionPassphrase=fnebJi6XYVoOIsvNDYO9KMec7ZI8qYbD3m0D0kYi5UJhoXh%2FviUkj7hqrvuPkXpaFgzSW0UqemP4JIOjFUMjmR2tWh%2Bt1DjRkr8uXnypprsGZvRoy7EGWdSwGemssovV; oc3kizzp7f7d=plgcob981he5sjek79r7avjqp1' \
    -H 'Content-Length:147' \
    -H 'Connection:Keep-Alive' \
    -H 'Accept-Encoding:gzip, deflate' \
    -H 'Accept-Language:en-US,*' \
    -H 'Host:alfageme.solidgear.prv:4443' \
    -X PROPFIND 'https://alfageme.solidgear.prv:4443/remote.php/webdav/' \
    --data-binary '<?xml version="1.0" ?>
<d:propfind xmlns:d="DAV:">
  <d:prop>
    <d:quota-available-bytes />
    <d:quota-used-bytes />
  </d:prop>
</d:propfind>
'
```

### iOS:

```bash
curl \
    -H 'Host:alfageme.solidgear.prv:4443' \
    -H 'Content-Type:application/xml' \
    -H 'Depth:1' \
    -H 'Accept:*/*' \
    -H 'Accept-Encoding:gzip, deflate' \
    -H 'Connection:keep-alive' \
    -H 'Cookie:oc3kizzp7f7d=6ds6rnc8mcl5jg31dkmdnpqli6; oc_sessionPassphrase=9sWpKyFrO8B59uXHo%2F4BRmHB3X2q6qJDtpS17D2Zx7BcFPp0s4yPdWiZBPqmzkmWvo4yoBW9XJ7%2BwzgtfCVWfmy%2F0TzJwWUVbROzy0LDGHFPM9543cKTkWnf6%2FqoFcB1' \
    -H 'User-Agent:Mozilla/5.0 (iOS) ownCloud-iOS/3.7.0' \
    -H 'Accept-Language:en-US;q=1, es-US;q=0.9, zh-Hant-US;q=0.8, en-GB;q=0.7, en;q=0.6' \
    -H 'Authorization:Bearer fVbLupKEkYi4Yr6tzALsAZlMZIjMDEkpjZXr5kcwYR6HMKvjDy0RQPoXS4dJ7QSl' \
    -H 'Content-Length:383' \
    -X PROPFIND 'https://alfageme.solidgear.prv:4443/remote.php/webdav/' \
    --data-binary '<?xml version="1.0" encoding="UTF-8"?><D:propfind xmlns:D="DAV:"><D:prop><D:resourcetype/><D:getlastmodified/><size xmlns="http://owncloud.org/ns"/><D:creationdate/><id xmlns="http://owncloud.org/ns"/><D:getcontentlength/><D:displayname/><D:quota-available-bytes/><D:getetag/><permissions xmlns="http://owncloud.org/ns"/><D:quota-used-bytes/><D:getcontenttype/></D:prop></D:propfind>'
```

### Diffs:

- Order of the `Authorization` and `Cookie` headers: shouldn't matter as WebDAV inherits HTTP's ["_The order in which header fields with differing field names are received is not significant_"](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2)
- Order of the `Cookie`s itself: `oc_sessionPassphrase` first/last
- `Depth`: 1/0
- Content of the request itself

Will play with these differences and report back when I find the one that messes with the auth. middleware.

**Comment by DeepDiver1975** [comment]
I honestly question if oauth works at all on the old endpoint .......

**Comment by PVince81** [comment]
If we didn't break more stuff that wasn't broken before I'd suggest moving forward with the PR and get it released as it adresses the more critical issue. Then investigate this detail separately ?

@SamuAlfageme @DeepDiver1975

**Comment by SamuAlfageme** [comment]
As discussed with @DeepDiver1975 and @owncloud/ios-developers, in the end, https://github.com/owncloud/oauth2/issues/103 was all about the iOS app circumventing the webview's sandbox (I haven't had the chance to check with Android's implementation, tho). This means the client was reusing cookies obtained during the authentication steps delegated to the browser (which goes against the end goal of the OAuth2 mechanism: isolation from this); those were bounded to the browser's activity; just like if they were obtained on safari or any other mobile browser, and valid as long as them (not dependent on the `access_token` at all).

After all the tests performed, all the basic access/refresh/revoke scenarios look sane enough. I'll go for the merge here. 👍

**Pull request merged**